// @wix/motion/dist/esm/library/mouse/Tilt3DMouse.js
import {
    mapRange
} from '@wix/wow-utils';
import {
    getMouseTransitionEasing
} from '../../utils';
import {
    BaseMouse
} from './BaseMouse';
class Tilt3DMouseAnimation extends BaseMouse {
    progress(_ref) {
        let {
            x: progressX,
            y: progressY
        } = _ref;
        const {
            inverted,
            angle,
            perspective
        } = this.options;
        const invert = inverted ? -1 : 1;

        // if progressX === 0, rotateX === -angle, if progressX === 0.5, rotateX === 0, if progressX === 1, rotateX === angle
        const rotateX = mapRange(0, 1, angle, -angle, progressY) * invert;
        const rotateY = mapRange(0, 1, -angle, angle, progressX) * invert;
        this.target.style.transform = `perspective(${perspective}px) rotateX(${rotateX}deg) rotateY(${rotateY}deg) rotate(var(--comp-rotate-z, 0deg))`;
    }
    cancel() {
        this.target.style.transform = '';
        this.target.style.transition = '';
    }
}
export default function create(options) {
    const {
        transitionDuration,
        transitionEasing
    } = options;
    const {
        inverted = false,
            angle = 5,
            perspective = 800
    } = options.namedEffect;
    const animationOptions = {
        transition: transitionDuration ? `transform ${transitionDuration}ms ${getMouseTransitionEasing(transitionEasing)}` : '',
        inverted,
        angle,
        perspective
    };
    return target => new Tilt3DMouseAnimation(target, animationOptions);
}
