// @wix/motion/dist/esm/library/backgroundScroll/BgCloseUp.js
export default function create(options) {
    const easing = 'linear';
    return [{
        ...options,
        easing,
        part: 'BG_MEDIA',
        startOffset: {
            name: 'cover',
            offset: {
                type: 'percentage',
                value: 50
            }
        },
        endOffset: {
            name: 'cover',
            offset: {
                type: 'percentage',
                value: 100
            }
        },
        keyframes: [{
            opacity: 1
        }, {
            opacity: 0
        }]
    }, {
        ...options,
        easing,
        part: 'BG_MEDIA',
        startOffset: {
            name: 'cover',
            offset: {
                type: 'percentage',
                value: 0
            }
        },
        endOffset: {
            name: 'cover',
            offset: {
                type: 'percentage',
                value: 100
            }
        },
        keyframes: [{
            transform: 'perspective(100px) translateZ(0px)'
        }, {
            transform: 'perspective(100px) translateZ(80px)'
        }]
    }];
    /*
     * @keyframes <name>-fade {
     *   from {
     *     opacity: <opacityFrom>;
     *   }
     *   to {
     *     opacity: <opacityTo>;
     *   }
     * }
     *
     * @keyframes <name>-zoom {
     *   from {
     *     transform: perspective(100px) translateZ(0px);
     *   }
     *   to {
     *     transform: perspective(100px) translateZ(80px);
     *   }
     * }
     */
}
